File: android/app/src/main/java/com/bradbumbalough/twilioipmessaging/RCTTwilioIPMessagingMessages.java
package com.bradbumbalough.twilioipmessaging;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;

import com.twilio.ipmessaging.Messages;
import com.twilio.ipmessaging.Constants;
import com.twilio.ipmessaging.ErrorInfo;
import com.twilio.ipmessaging.Message;

import java.util.List;

public class RCTTwilioIPMessagingMessages extends ReactContextBaseJavaModule {

    @Override
    public String getName() {
        return "TwilioIPMessagingMembers";
    }

    private ReactApplicationContext reactContext;

    public RCTTwilioIPMessagingMessages(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    private Messages loadMessagesFromChannelSid(String sid) {
        return RCTTwilioIPMessagingClient.getInstance().client.getChannels().getChannel(sid).getMessages();
    }

    @ReactMethod
    public void getLastConsumedMessageIndex(String channelSid, Promise promise) {
        promise.resolve(loadMessagesFromChannelSid(channelSid).getLastConsumedMessageIndex());
    }

    @ReactMethod
    public void sendMessage(String channelSid, String body, final Promise promise) {
        Messages messages = loadMessagesFromChannelSid(channelSid);

        Constants.StatusListener listener = new Constants.StatusListener() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("send-message-error","Error occurred while attempting to sendMessage.");
            }

            @Override
            public void onSuccess() {
                promise.resolve(true);
            }
        };

        messages.sendMessage(messages.createMessage(body), listener);
    }

    @ReactMethod
    public void removeMessage(String channelSid, Integer index, final Promise promise) {
        Messages messages = loadMessagesFromChannelSid(channelSid);

        Message messageToRemove = null;
        try {
            for (Message m : loadMessagesFromChannelSid(channelSid).getMessages()) {
                if (m.getMessageIndex() == index) {
                    messageToRemove = m;
                    break;
                }
            }
        }
        catch (Exception e) {
            promise.reject("remove-message-error","Error occurred while attempting to remove message.");
        }

        Constants.StatusListener listener = new Constants.StatusListener() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("remove-message-error","Error occurred while attempting to remove message.");
            }

            @Override
            public void onSuccess() {
                promise.resolve(true);
            }
        };

        loadMessagesFromChannelSid(channelSid).removeMessage(messageToRemove, listener);
    }

    @ReactMethod
    public void getLastMessages(String channelSid, Integer count, final Promise promise) {

        Constants.CallbackListener<List<Message>> listener = new Constants.CallbackListener<List<Message>>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("get-last-message-error","Error occurred while attempting to getLastMessage.");
            }

            public void onSuccess(List<Message> messages) {
                promise.resolve(RCTConvert.Messages(messages));
            }
        };

        loadMessagesFromChannelSid(channelSid).getLastMessages(count, listener);
    }

    @ReactMethod
    public void getMessagesAfter(String channelSid, Integer index, Integer count, final Promise promise) {

        Constants.CallbackListener<List<Message>> listener = new Constants.CallbackListener<List<Message>>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("get-messages-after-error","Error occurred while attempting to getMessagesAfter.");
            }

            public void onSuccess(List<Message> messages) {
                promise.resolve(RCTConvert.Messages(messages));
            }
        };

        loadMessagesFromChannelSid(channelSid).getMessagesAfter(index, count, listener);
    }

    @ReactMethod
    public void getMessagesBefore(String channelSid, Integer index, Integer count, final Promise promise) {

        Constants.CallbackListener<List<Message>> listener = new Constants.CallbackListener<List<Message>>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("get-messages-before-error","Error occurred while attempting to getMessagesBefore.");
            }

            public void onSuccess(List<Message> messages) {
                promise.resolve(RCTConvert.Messages(messages));
            }
        };

        loadMessagesFromChannelSid(channelSid).getMessagesBefore(index, count, listener);
    }

    @ReactMethod
    public void getMessage(String channelSid, Integer index, final Promise promise) {
        Constants.CallbackListener<List<Message>> listener = new Constants.CallbackListener<List<Message>>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("get-message-error","Error occurred while attempting to getMessage.");
            }

            public void onSuccess(List<Message> messages) {
                promise.resolve(RCTConvert.Message(messages.get(0)));
            }
        };

        loadMessagesFromChannelSid(channelSid).getMessagesAfter(index, 1, listener);
    }

    @ReactMethod
    public void setLastConsumedMessageIndex(String channelSid, Integer index) {
        loadMessagesFromChannelSid(channelSid).setLastConsumedMessageIndex(index);
    }

    @ReactMethod
    public void advanceLastConsumedMessageIndex(String channelSid, Integer index) {
        loadMessagesFromChannelSid(channelSid).advanceLastConsumedMessageIndex(index);
    }

    @ReactMethod
    public void setAllMessagesConsumed(String channelSid) {
        loadMessagesFromChannelSid(channelSid).setAllMessagesConsumed();
    }

    // Message instance method

    @ReactMethod
    public void updateBody(String channelSid, Integer index, String body, final Promise promise) {;

        Constants.StatusListener listener = new Constants.StatusListener() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("update-body-error","Error occurred while attempting to updateBody.");
            }

            public void onSuccess() {
                promise.resolve(true);
            }
        };

        try {
            for (Message m : loadMessagesFromChannelSid(channelSid).getMessages()) {
                if (m.getMessageIndex() == index) {
                    m.updateMessageBody(body, listener);
                    break;
                }
            }
        }
        catch (Exception e) {
            promise.reject("update-body-error","Error occurred while attempting to updateBody.");
        }
    }

}

File: ios/RCTTwilioIPMessaging/RCTTwilioIPMessagingMessages.m
//
//  RCTTwilioIPMessagingMessages.m
//  TwilioIPExample
//
//  Created by Brad Bumbalough on 6/3/16.
//  Copyright © 2016 Facebook. All rights reserved.
//

#import "RCTTwilioIPMessagingMessages.h"
#import "RCTTwilioIPMessagingChannels.h"
#import "RCTTwilioIPMessagingClient.h"
#import "RCTConvert+TwilioIPMessagingClient.h"
#import <RCTUtils.h>

@implementation RCTTwilioIPMessagingMessages

RCT_EXPORT_MODULE()

- (TWMMessages *)loadMessagesFromChannelSid:(NSString *)channelSid {
  TwilioIPMessagingClient *client = [[RCTTwilioIPMessagingClient sharedManager] client];
  return [[[client channelsList] channelWithId:channelSid] messages];
}

#pragma mark Messages Methods

RCT_REMAP_METHOD(getLastConsumedMessageIndex, channelSid:(NSString *)channelSid last_consumed_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  TWMMessages *messages = [self loadMessagesFromChannelSid:channelSid];
  resolve(@[RCTNullIfNil(messages.lastConsumedMessageIndex)]);
}


RCT_REMAP_METHOD(sendMessage, channelSid:(NSString *)channelSid body:(NSString *)body send_message_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  TWMMessages *messages = [self loadMessagesFromChannelSid:channelSid];
  [messages sendMessage:[messages createMessageWithBody:body] completion:^(TWMResult *result) {
    if (result.isSuccessful) {
      resolve(@[@TRUE]);
    }
    else {
      reject(@"send-message-error", @"error", result.error);
    }
  }];
}

RCT_REMAP_METHOD(removeMessage, channelSid:(NSString *)channelSid index:(NSNumber *)index remove_message_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  TWMMessages *messages = [self loadMessagesFromChannelSid:channelSid];
  [messages removeMessage:[messages messageWithIndex:index] completion:^(TWMResult *result) {
    if (result.isSuccessful) {
      resolve(@[@TRUE]);
    }
    else {
      reject(@"remove-message-error", @"error", result.error);
    }
  }];
}

RCT_REMAP_METHOD(getLastMessages, channelSid:(NSString *)channelSid count:(NSUInteger)count get_message_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  TWMMessages *messages = [self loadMessagesFromChannelSid:channelSid];
  [messages getLastMessagesWithCount:count completion:^(TWMResult *result, NSArray<TWMMessage *> *messages) {
    if (result.isSuccessful) {
      resolve([RCTConvert TWMMessages:messages]);
    }
    else {
      reject(@"get-message-error", @"error", result.error);
    }
  }];
}

RCT_REMAP_METHOD(getMessagesBefore, channelSid:(NSString *)channelSid index:(NSUInteger)index withCount:(NSUInteger)withCount get_message_before_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  TWMMessages *messages = [self loadMessagesFromChannelSid:channelSid];
  [messages getMessagesBefore:index withCount:withCount completion:^(TWMResult *result, NSArray<TWMMessage *> *messages) {
    if (result.isSuccessful) {
      resolve([RCTConvert TWMMessages:messages]);
    }
    else {
      reject(@"get-message-error", @"error", result.error);
    }
  }];
}

RCT_REMAP_METHOD(getMessagesAfter, channelSid:(NSString *)channelSid index:(NSUInteger)index withCount:(NSUInteger)withCount get_message_after_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  TWMMessages *messages = [self loadMessagesFromChannelSid:channelSid];
  [messages getMessagesAfter:index withCount:withCount completion:^(TWMResult *result, NSArray<TWMMessage *> *messages) {
    if (result.isSuccessful) {
      resolve([RCTConvert TWMMessages:messages]);
    }
    else {
      reject(@"get-message-error", @"error", result.error);
    }
  }];
}

RCT_REMAP_METHOD(getMessage, channelSid:(NSString *)channelSid index:(NSNumber *)index message_index_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  TWMMessages *messages = [self loadMessagesFromChannelSid:channelSid];
  resolve([RCTConvert TWMMessage:[messages messageWithIndex:index]]);
}

RCT_REMAP_METHOD(messageForConsumptionIndex, channelSid:(NSString *)channelSid index:(NSNumber *)index consumption_message_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    TWMMessages *messages = [self loadMessagesFromChannelSid:channelSid];
    resolve([RCTConvert TWMMessage:[messages messageForConsumptionIndex:index]]);
}

RCT_REMAP_METHOD(setLastConsumedMessageIndex, channelSid:(NSString *)channelSid set_index:(NSNumber *)index) {
  TWMMessages *messages = [self loadMessagesFromChannelSid:channelSid];
  [messages setLastConsumedMessageIndex:index];
}

RCT_REMAP_METHOD(advanceLastConsumedMessageIndex, channelSid:(NSString *)channelSid advance_index:(NSNumber *)index) {
  TWMMessages *messages = [self loadMessagesFromChannelSid:channelSid];
  [messages advanceLastConsumedMessageIndex:index];
}

RCT_REMAP_METHOD(setAllMessagesConsumed, channelSid:(NSString *)channelSid) {
  TWMMessages *messages = [self loadMessagesFromChannelSid:channelSid];
  [messages setAllMessagesConsumed];
}

#pragma mark Message Instance Methods

RCT_REMAP_METHOD(updateBody, channelSid:(NSString *)channelSid index:(NSNumber *)index body:(NSString *)body update_message_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  TWMMessages *messages = [self loadMessagesFromChannelSid:channelSid];
  [[messages messageWithIndex:index] updateBody:body completion:^(TWMResult *result) {
    if (result.isSuccessful) {
      resolve(@[@TRUE]);
    }
    else {
      reject(@"update-message-error", @"error", result.error);
    }
  }];
}

@end
